docs/protocol/core/v3.0.rst

```diff
@@ -728,17 +728,221 @@ For example, the JSON document below defines a group::
+from a Zarr hierarchy. For a store to be compatible with this
+protocol, it must support a set of operations defined in the `Store
+interface`_ subsection. The store interface can be implemented using a
+variety of underlying storage technologies, described in the
+subsection on `Store implementations`_.
+
 Store interface
----------------
+~~~~~~~~~~~~~~~
+
+The store interface is intended to be simple to implement using a
+variety of different underlying storage technologies. It is defined in
+a general way here, but it should be straightforward to translate into
+a software interface in any given programming language. The goal is
+that an implementation of this specification could be modular and
+allow for different store implementations to be used.
+
+The store interface defines a set of operations involving `keys` and
+`values`. In the context of this interface, a `key` is any ASCII
+string containing only characters in the ranges ``a-z``, ``A-Z``,
+``0-9``, or in the set ``/.-_`` and a `value` is any sequence of
+bytes. It is assumed that the store holds (`key`, `value`) pairs, with
+only one such pair for any given `key`. I.e., a store is a mapping
+from keys to values.
+
+The following operations are **required**:
+
+``get`` - Retrieve the `value` associated with a given `key`.
+
+    | Parameters: `key`
+    | Output: `value`
+
+``set`` - Store a (`key`, `value`) pair.
+
+    | Parameters: `key`, `value`
+    | Output: none
+
+``delete`` - Delete the given key/value pair from the store.
+
+    | Parameters: `key`
+    | Output: none
+
+The following operations are **optional**, but if supported may be
+used by an implementation of this protocol to provide additional
+functionality and/or optimisations:
+
+``list`` - Retrieve all `keys` in the store.
 
-TODO define the store interface
+    | Parameters: none
+    | Output: set of `keys`
+
+``listpre`` - Retrieve all keys with a given prefix.
+
+    | Parameters: `prefix`
+    | Output: set of `keys` with the given `prefix`
+
+    For example, if a store contains the keys "a/b", "a/c/d" and
+    "e/f/g", then ``listpre("a/")`` would return "a/b" and "a/c/d".
+
+    If ``listpre`` is not supported by a store, an implementation of
+    this protocol could fall back to ``list`` with post-processing of
+    the results.
+
+``listdir`` - Retrieve all keys and prefixes with a given prefix and
+which do not contain "/" after the given prefix.
+
+    | Parameters: `prefix`
+    | Output: set of `keys` and set of `prefixes`
+
+    For example, if a store contains the keys "a/b", "a/c", "a/d/e",
+    "a/f/g", then ``listdir("a/")`` would return keys "a/b" and "a/c"
+    and prefixes "a/d/" and "a/f/".
+
+    If ``listdir`` is not supported by a store, an implementation of
+    this protocol could fallback to ``listpre`` with post-processing
+    of the results.
+
+
+Store implementations
+~~~~~~~~~~~~~~~~~~~~~
+
+(This subsection is not normative.)
+
+TODO describe possible implementations.
```

Comment:
A python dict? Probably the simplest, except the methods are not named as above

Reply:
Yes that would be a good example. I was thinking to use this subsection just to describe briefly a small variety of different possible implementations, to reinforce that this protocol is not just about file system storage.